Framework/MDAlgorithms/src/IntegrateEllipsoidsTwoStep.cpp

```diff
@@ -0,0 +1,553 @@
+                  "different integration radius.  Q includes the 2*pi factor.");
+
+  declareProperty("WeakPeakThreshold", 1.0, mustBePositive,
+                  "Intensity threshold use to classify a peak as weak.");
+
+  declareProperty(
+      make_unique<WorkspaceProperty<PeaksWorkspace>>("OutputWorkspace", "",
+                                                     Direction::Output),
+      "The output PeaksWorkspace will be a copy of the input PeaksWorkspace "
+      "with the peaks' integrated intensities.");
+}
+
+void IntegrateEllipsoidsTwoStep::exec() {
+  PeaksWorkspace_sptr input_peak_ws = getProperty("PeaksWorkspace");
+  MatrixWorkspace_sptr input_ws = getProperty("InputWorkspace");
+  EventWorkspace_sptr eventWS =
+      boost::dynamic_pointer_cast<EventWorkspace>(input_ws);
+
+  Workspace2D_sptr histoWS = boost::dynamic_pointer_cast<Workspace2D>(input_ws);
+  if (!eventWS && !histoWS) {
+    throw std::runtime_error("IntegrateEllipsoids needs either a "
+                             "EventWorkspace or Workspace2D as input.");
+  }
+
+  const double weakPeakThreshold = getProperty("WeakPeakThreshold");
+
+  // validation of inputs
+  if (!input_peak_ws) {
+    throw std::runtime_error("Could not read the Peaks Workspace");
+  }
+
+  if (!input_ws) {
+    throw std::runtime_error("Could not read the Input Workspace");
+  }
+
+  PeaksWorkspace_sptr peak_ws = getProperty("OutputWorkspace");
+  if (peak_ws != input_peak_ws) {
+    peak_ws = input_peak_ws->clone();
+  }
+
+  Progress prog(this, 0.5, 1.0, input_ws->getNumberHistograms());
+
+  std::vector<Peak> &peaks = peak_ws->getPeaks();
+  size_t n_peaks = peak_ws->getNumberPeaks();
+  size_t indexed_count = 0;
+  std::vector<V3D> peak_q_list;
+  std::vector<V3D> hkl_vectors;
+  for (size_t i = 0; i < n_peaks; i++) // Note: we skip un-indexed peaks
+  {
+    V3D hkl(peaks[i].getH(), peaks[i].getK(), peaks[i].getL());
+    if (Geometry::IndexingUtils::ValidIndex(hkl, 1.0)) // use tolerance == 1 to
+                                                       // just check for (0,0,0)
+    {
+      peak_q_list.emplace_back(peaks[i].getQLabFrame());
+      V3D miller_ind(static_cast<double>(boost::math::iround<double>(hkl[0])),
+                     static_cast<double>(boost::math::iround<double>(hkl[1])),
+                     static_cast<double>(boost::math::iround<double>(hkl[2])));
+      hkl_vectors.push_back(miller_ind);
+      indexed_count++;
+    }
+  }
+
+  if (indexed_count < 3) {
+    throw std::runtime_error(
+        "At least three linearly independent indexed peaks are needed.");
+  }
+  // Get UB using indexed peaks and
+  // lab-Q vectors
+  Matrix<double> UB(3, 3, false);
+  Geometry::IndexingUtils::Optimize_UB(UB, hkl_vectors, peak_q_list);
+  Matrix<double> UBinv(UB);
+  UBinv.Invert();
+  UBinv *= (1.0 / (2.0 * M_PI));
+
+  std::vector<std::pair<double, V3D>> qList;
+  for (size_t i = 0; i < n_peaks; i++) {
+    qList.emplace_back(1., V3D(peaks[i].getQLabFrame()));
```

Comment:
@AnthonyLim23 Why `push_back` over `emplace_back`? `emplace` allows the new object to be constructed in place.